### ISSUE TITLE
Molecule Initial Setup

### ISSUE THREAD

**Issue by mew2057** [IBM/ibm-spectrum-scale-csi-operator#36]
Molecule needs to be initialized and green for before we can start writing additional tests. Right now there's a failure relating to missing variables:

```
TASK [csi-scale : Ensure scale_config] *****************************************

    fatal: [localhost]: FAILED! => {"msg": "The task includes an option with an undefined variable. The error was: 'dict object' has no attribute 'spec'\n\nThe error appears to be in '/home/travis/build/mew2057/ibm-spectrum-scale-csi-operator/stable/ibm-spectrum-scale-csi-operator-bundle/operators/ibm-spectrum-scale-csi-operator/roles/csi-scale/tasks/plugin_config.yml': line 2, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n---\n- name: \"Ensure scale_config\"\n  ^ here\n"}
```

This is likely indicative that we're handling variables wrong in the role. 

This issue is a tracking issue for getting molecule to the `green` state.

**Comment by mew2057** [comment]
- [ ] Remove `_scale_ibm_com_csiscaleoperator` from the task file. It places too much dependence on the operator framework and reduces the overall flexibility of the operator.

- [ ] Add sensible defaults for variables that may be missing.

**Comment by mew2057** [comment]
It looks like `camelCase` variables aren't put into the `vars` for the ansible playbook:

``` YAML

apiVersion: scale.ibm.com/v1alpha1
kind: 'CSIScaleOperator'
metadata:
    # IMPORTANT THE OPERATOR LOOKS FOR THIS!
    name: 'csi-scale-operator'
status: {}
spec:
  #
  scaleHostpath: "/ibm/fs1"
  secretCounter: 1000
  attacher: "test123"
  provisioner: "test23"
  driverRegistrar: "test3"
  spectrumScale: "test"

  clusters:
    - id: "11832033572270148010"
      secrets: "secret1"
      secureSslMode: false
      primary:
        primaryFS: "fs1"
        primaryFset: "csiFset2"
      restApi:
      - guiHost: "jdunham-rh72-k8s-scale-master.fyre.ibm.com"
```

`attacher`, `provisioner` and `cluster` are all processed, but  `scaleHostpath`, `secretCounter, `driverRegistrar` and `spectrumScale` are all set to their defaults.
``` bash
TASK [csi-scale : debug] *******************************************************
task path: /opt/ansible/roles/csi-scale/tasks/main.yml:9
ok: [localhost] => {
    "msg": [
        "scaleHostpath - /",
        "secretCounter - -1",
        "attacher - test123",
        "provisioner - test23",
        "driverRegistrar -quay.io/k8scsi/csi-node-driver-registrar:v1.0.1",
        "spectrumScale - quay.io/mew2057/ibm-spectrum-scale-csi-driver:v0.9.0"
    ]
}
META: ending play
```

**Comment by mew2057** [comment]
Looked into it and lo and behold the operator changes the case:
```
5150             "driverRegistrar": "quay.io/k8scsi/csi-node-driver-registrar:v1.0.1",^M
5151             "driver_registrar": "test3",^M
```

**Comment by mew2057** [comment]
``` YAML
## defaults//main.yml
---
# defaults file for ansible-role-csi-scale
name: "csi-spectrum-scale"
namespace: '{{ meta.namespace }}'
storage-class: "csi-spectrum-scale"


# Image defaults
attacher:    "quay.io/k8scsi/csi-attacher:v1.0.0"
provisioner: "quay.io/k8scsi/csi-provisioner:v1.0.0"

# Due to camelCase these get loaded differently in the operator.
driverRegistrar: "{{ driver_registrar | default('quay.io/k8scsi/csi-node-driver-registrar:v1.0.1') }}"
spectrumScale:   "{{ spectrum_scale   | default('quay.io/mew2057/ibm-spectrum-scale-csi-driver:v0.9.0') }}"

# Set defaults for the secret counter and hostpath.
secretCounter: "{{ secret_counter |  default(-1) }}"
scaleHostpath: "{{ scale_hostpath | default('/') }}"




# State of driver in the cluster (might not be needed).
#state: absent
state: present
```

**Comment by mew2057** [comment]
That looks like it fixed that issue, now onto the next one.﻿
``` bash
TASK [csi-scale : debug] *******************************************************
task path: /opt/ansible/roles/csi-scale/tasks/main.yml:9
ok: [localhost] => {
    "msg": [
        "scaleHostpath - /ibm/fs1",
        "secretCounter - 1001",
        "attacher - test123",
        "provisioner - test23",
        "driverRegistrar -test3",
        "spectrumScale - test"
    ]
}
META: ending play
```

**Comment by mew2057** [comment]
``` bash
[root csi-scale]# kubectl logs csi-spectrum-scale-w6f4s csi-spectrum-scale
ERROR: logging before flag.Parse: I1030 19:13:41.532922       1 gpfs.go:108] scale: Loaded 0 volumes from /var/lib/kubelet/plugins/csi-spectrum-scale/controller
I1030 19:13:41.533544       1 gpfs.go:112] gpfs GetScaleDriver
I1030 19:13:41.534648       1 gpfs.go:189] gpfs SetupScaleDriver. name: csi-spectrum-scale, version: 1.0.0, nodeID: worker1
I1030 19:13:41.534699       1 gpfs.go:228] gpfs PluginInitialize
I1030 19:13:41.534961       1 scale_config.go:74] scale_config LoadScaleConfigSettings
I1030 19:13:41.536603       1 scale_config.go:97] scale_config HandleSecrets
I1030 19:13:41.536701       1 gpfs.go:430] gpfs ValidateScaleConfigParameters.
E1030 19:13:41.536715       1 gpfs.go:233] Parameter validation failure
E1030 19:13:41.536720       1 gpfs.go:196] Error in plugin initialization: Mandatory parameters not specified for cluster 11832033572270148010
F1030 19:13:41.536733       1 main.go:65] Failed to initialize Scale CSI Driver: Mandatory parameters not specified for cluster 11832033572270148010
```

Looks like the thing that changes the case  hits the clusters as well? 

Trying a new prop loader for clusters:
``` YAML
clusters: "{{ _scale_ibm_com_csiscaleoperator.spec.clusters | default([])}}"
```

**Comment by mew2057** [comment]
Looks like the snake_case is documented here https://docs.openshift.com/container-platform/4.1/applications/operator_sdk/osdk-ansible.html#osdk-building-ansible-operator_osdk-ansible

**Comment by mew2057** [comment]
Latest error is that the molecule framework can't find the template files. Need to investigate:
``` bash
TASK [csi-scale : Ensure csi-scale objects are present] ************************
    failed: [localhost] (item={'name': 'spectrum_scale.yaml.j2'}) => {"ansible_loop_var": "item", "changed": false, "error": 404, "item": {"name": "spectrum_scale.yaml.j2"}, "msg": "Failed to create object: b'{\"kind\":\"Status\",\"apiVersion\":\"v1\",\"metadata\":{},\"status\":\"Failure\",\"message\":\"namespaces \\\\\"osdk-test\\\\\" not found\",\"reason\":\"NotFound\",\"details\":{\"name\":\"osdk-test\",\"kind\":\"namespaces\"},\"code\":404}\\n'", "reason": "Not Found", "status": 404}
    failed: [localhost] (item={'name': 'csi-plugin-attacher.yaml.j2'}) => {"ansible_loop_var": "item", "changed": false, "error": 404, "item": {"name": "csi-plugin-attacher.yaml.j2"}, "msg": "Failed to create object: b'{\"kind\":\"Status\",\"apiVersion\":\"v1\",\"metadata\":{},\"status\":\"Failure\",\"message\":\"namespaces \\\\\"osdk-test\\\\\" not found\",\"reason\":\"NotFound\",\"details\":{\"name\":\"osdk-test\",\"kind\":\"namespaces\"},\"code\":404}\\n'", "reason": "Not Found", "status": 404}
    failed: [localhost] (item={'name': 'csi-plugin-provisioner.yaml.j2'}) => {"ansible_loop_var": "item", "changed": false, "error": 404, "item": {"name": "csi-plugin-provisioner.yaml.j2"}, "msg": "Failed to create object: b'{\"kind\":\"Status\",\"apiVersion\":\"v1\",\"metadata\":{},\"status\":\"Failure\",\"message\":\"namespaces \\\\\"osdk-test\\\\\" not found\",\"reason\":\"NotFound\",\"details\":{\"name\":\"osdk-test\",\"kind\":\"namespaces\"},\"code\":404}\\n'", "reason": "Not Found", "status": 404}
    failed: [localhost] (item={'name': 'csi-plugin.yaml.j2'}) => {"ansible_loop_var": "item", "changed": false, "error": 404, "item": {"name": "csi-plugin.yaml.j2"}, "msg": "Failed to create object: b'{\"kind\":\"Status\",\"apiVersion\":\"v1\",\"metadata\":{},\"status\":\"Failure\",\"message\":\"namespaces \\\\\"osdk-test\\\\\" not found\",\"reason\":\"NotFound\",\"details\":{\"name\":\"osdk-test\",\"kind\":\"namespaces\"},\"code\":404}\\n'", "reason": "Not Found", "status": 404}
```

**Comment by mew2057** [comment]
Nice, got the build working.

https://travis-ci.org/mew2057/ibm-spectrum-scale-csi-operator/jobs/605617257

**Comment by mew2057** [comment]
Got staging with parallel workloads:
https://travis-ci.org/mew2057/ibm-spectrum-scale-csi-operator/jobs/605652589?utm_medium=notification&utm_source=email

**Comment by mew2057** [comment]
While the namespace fix worked nicely for the molecule test, it broke the operator (whoops). Which tells us the tests are not yet fully up to snuff.

Also molecule test points to a bad habit of using `name` in the roles. Renaming name to `opName`.

**Comment by mew2057** [comment]
molecule tests are down for tonight, still trying to get a greenlight on the namespace code.

**Comment by mew2057** [comment]
Figured out why my tests weren't running:

**Comment by mew2057** [comment]
Got a green build https://travis-ci.org/mew2057/ibm-spectrum-scale-csi-operator/jobs/606101356

I need to do more optimization, however, because it's taking 20 mins to build (whoops).

**Comment by mew2057** [comment]
Locally `molecule test -s test-local` doesn't actually work still, but I think this is a configuration issue more than anything else:

``` bash
    TASK [Build Operator Image] ****************************************************
    fatal: [kind-test-local]: FAILED! => {"changed": true, "cmd": ["docker", "build", "-f", "/build/build/Dockerfile", "-t", "ibm.scale.com/csi-scale-operator:testing", "/build"], "delta": "0:00:00.763002", "end": "2019-11-01 17:33:53.079117", "msg": "non-zero return code", "rc": 1, "start": "2019-11-01 17:33:52.316115", "stderr": "Get https://quay.io/v2/: x509: certificate is valid for jdunham-rh72-k8s-scale-master.fyre.ibm.com, not quay.io", "stderr_lines": ["Get https://quay.io/v2/: x509: certificate is valid for jdunham-rh72-k8s-scale-master.fyre.ibm.com, not quay.io"], "stdout": "Sending build context to Docker daemon  44.11MB\r\r\nStep 1/4 : FROM quay.io/operator-framework/ansible-operator:v0.11.0", "stdout_lines": ["Sending build context to Docker daemon  44.11MB", "", "Step 1/4 : FROM quay.io/operator-framework/ansible-operator:v0.11.0"]}
```
Based on this error it looks like build is failing due to a certificate error.

**Comment by mew2057** [comment]
Okay, so parallel vs serial is basically the same:
![Screen Shot 2019-11-01 at 1 43 54 PM](https://user-images.githubusercontent.com/1195452/68044433-b48a1200-fcad-11e9-9009-4121fdf28f01.png)

**Comment by mew2057** [comment]
Running the `test-cluster`, somehow the deployment never gets made?

**Comment by mew2057** [comment]
Since the test-local works I'm freezing additional work pending an offline meeting Monday (results will be posted  here).

**Comment by mew2057** [comment]
These changes have been merged Issue  #47 is now tracking molecule enhancements.
